File: c++/heap/twoHeaps.cpp
#include <functional> // lambda functions
#include <vector>
#include <algorithm> // for the heap (not a container but and algorithms like the std::sort)
#include <iostream>

using namespace std;

/**
 * Documentation for heap algorithms: https://www.geeksforgeeks.org/cpp-stl-heap/
 * The following is an implementation of two heaps for finding the min value in a stream of numbers.
    Brute force apprach: sort the seqeunce every time and return the median everytime.
     - This would be a O(n^2 log(n)) big omega approach if using mergesort and a O(n^3) big O approach if using quick sort.
    Our approach: using two different heaps. A small max heap and a large min heap
 */

class MedianFinder
{
public:
    MedianFinder() {}

    void addNum(int num)
    {
        if (large.size() > 0 && num > large.front())
        {
            large.push_back(num);
            push_heap(large.begin(), large.end(), minComp);
        }
        else
        {
            small.push_back(num);
            push_heap(small.begin(), small.end(), maxComp);
        }

        if (small.size() > large.size() + 1)
        {
            int largest_small_val = small.front();
            pop_heap(small.begin(), small.end(), maxComp); // move largest to the back
            small.pop_back();
            // add the largest val to the large heap
            large.push_back(largest_small_val);
            push_heap(large.begin(), large.end(), minComp);
        }
        else if (large.size() > small.size() + 1)
        {
            int smallest_large_val = large.front();
            pop_heap(large.begin(), large.end(), minComp);
            large.pop_back();
            small.push_back(smallest_large_val);
            push_heap(small.begin(), small.end(), maxComp);
        }
    }

    double findMedian()
    {
        cout << small.size() << large.size();
        if (small.size() > large.size())
            return small.front();
        else if (small.size() < large.size())
            return large.front();
        return (float(small.front() + large.front()) / 2);
    }

private:
    vector<int> small; // min heap
    vector<int> large; // max heap

    // declaring comparator
    // static bool minComp(int a, int b){ // this is needed so that the other classes can access (no this pointer)
    //     return a > b;
    // }

    // approach using std lib functions (lambda functions)
    function<bool(int, int)> minComp = [](int a, int b)
    { return a > b; };
    function<bool(int, int)> maxComp = [](int a, int b)
    { return a < b; };

    // apprach three, functors
    // struct comp {
    //     bool operator()(int a, int b) const {
    //         return a > b;
    //     }
    // };

    // comp minComp;
};

File: Python/Heap/TwoHeaps.py

import heapq

""" The following is an implementation of two heaps for finding the min value in a stream of numbers.
    Brute force apprach: sort the seqeunce every time and return the median everytime.
     - This would be a O(n^2 log(n)) big omega approach if using mergesort and a O(n^3) big O approach if using quick sort.
    Our approach: using two different heaps. A small max heap and a large min heap."""

class MedianFinder:

    def __init__(self):
        self.small = []
        self.large = []

    def addNum(self, num: int) -> None:

        # push into appropriate heap
        if self.large and num > self.large[0]:
            heapq.heappush(self.large, num)
        else:
            heapq.heappush(self.small, num * -1)
        # heapq.heappush(self.small, num * -1)
        # if (self.large and self.small[0] * -1 > self.large[0]):
        #     val = heapq.heappop(self.small)
        #     heapq.heappush(self.large, val * -1)
        
        # handle case where there's a size mismatch
        if (len(self.small) > len(self.large) + 1):
            val = heapq.heappop(self.small)
            heapq.heappush(self.large, val * -1)
        elif (len(self.small) + 1 < len(self.large)):
            val = heapq.heappop(self.large)
            heapq.heappush(self.small, val * -1)


    def findMedian(self) -> float:
        if (len(self.small) > len(self.large)):
            return self.small[0] * -1
        elif (len(self.small) < len(self.large)):
            return self.large[0]
        return ((self.small[0] * -1) + (self.large[0])) / 2
        